Paletted PNG output

jblarsen opened the pull request:
Hi Rob,

I have made an option to convert the resulting PNG to a paletted PNG. The conversion is only done when the option is enabled and when the number of colors in the plot is <= 256. I have tested the conversion against the coads.nc example file (and our own files) using this script:

http://pastebin.com/ZnXSSaJw

You should be able to run this yourself. Just run the script from a directory where you have coads.nc (the script will store some pngs in this directory). On our server I get the following results:

$ python test_paletted_png.py
Timing png generation. Please be patient...
Average cost of paletting one png: 23.85 milliseconds
Total size reduced from 102688 to 58788 (42.8 %)

It depends on the specific use case whether such a conversion makes sense (whether the extra time it takes to do the paletting outweighs the time saved transmitting the image - taking into account that the same image can be cached and sent to multiple users). For our purposes we definitely expect it to be worth the extra CPU time. Especially for quiver plots (which I have an unfinished test implementation of - I will of course submit it so that you can decide whether you want it as well when it is more polished). But your mileage might vary (which is also why it should IMO be optional).

One assumption of this conversion is that we only have GIF-style transparency. One pixel is either fully transparent or fully opaque. But that is of course also the case for the plots that are produced.

Best regards,
Jesper

Commit:
added optional paletted png output

Commit:
safe bool casting

Commit:
only convert for ncolors<=256

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
make PIL optional

betodealmeida commented on pydap/responses/wms/__init__.py:
```diff
@@ -19,6 +19,10 @@
+try:
+    from PIL import Image
+except:
+    pass
```
I would change this to `Image = None` instead of `pass`, so we can test if PIL was successfully imported.

betodealmeida commented on pydap/responses/wms/__init__.py:
```diff
@@ -233,7 +237,32 @@ def serialize(dataset):
-            canvas.print_png(output)
+            # Optionally convert to paletted png
+            paletted = asbool(environ.get('pydap.responses.wms.paletted', 'false'))
+            if paletted:
```
And here replace with `if paletted and Image:`, this way the code won't fail if the variable is set but PIL is not installed.

betodealmeida commented:
We should also add PIL as a dependency to `pavement.py`. Have you thought of using `Pillow` instead?

Commit:
bugfix: im.getcolors() returns None for > 256 colors

Commit:
added Pillow as a requirement

jblarsen commented:
Hi Rob,

Thanks for the quick feedback. I have fixed a bug and changed the import
statement so that PIL is None when the module is not present. This new
version has been pushed to github. If it is None should I then silently
ignore the paletting, throw an exception or something else?

I am actually using Pillow. But it is a drop-in replacement for PIL even
taking PILs module name. I forgot to include it as a dependency in
pavement.py. This is also done and pushed to github.

Best regards,
Jesper

2014-03-20 0:07 GMT+01:00 Roberto De Almeida notifications@github.com:

> We should also add PIL as a dependency to pavement.py. Have you thought
> of using Pillow instead?
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/robertodealmeida/pydap.responses.wms/pull/2#issuecomment-38118621
> .
